interp/interp_test.go

```diff
@@ -3943,6 +3964,49 @@ func TestRunnerContext(t *testing.T) {
 	}
 }
 
+func TestCancelreader(t *testing.T) {
+	t.Parallel()
+
+	p := syntax.NewParser()
+	file := parse(t, p, "read x")
+	ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
+	// Make the linter happy, even though we deliberately wait for the
+	// timeout.
+	defer cancel()
+
+	var r *interp.Runner
+	if runtime.GOOS == "windows" {
+		// On Windows, the cancelreader only works on stdin
+		r, _ = interp.New(interp.StdIO(os.Stdin, nil, nil))
+	} else {
+		siR, siW, err := os.Pipe()
+		if err != nil {
+			t.Fatalf("Error calling os.Pipe: %v", err)
+		}
+		defer func() {
+			siW.Close()
+			siR.Close()
+		}()
+		r, _ = interp.New(interp.StdIO(siR, nil, nil))
```

Comment:
nit: move interp.New and StdIO outside the if/else, you can just have each branch set the stdin file var